### PR TITLE
Allow creating public knock rooms

### PR DESCRIPTION
**Pull request by charlynguyen**
We @nordeck are currently implementing the knock rooms behind the feature flag `feature_ask_to_join` (introduced in https://github.com/matrix-org/matrix-react-sdk/pull/11182).

This pull request allows creating public knock rooms.

Follow up of: https://github.com/matrix-org/matrix-react-sdk/pull/11464
Epic: https://github.com/vector-im/element-web/issues/18655

#### Relates to

- https://github.com/matrix-org/matrix-react-sdk/pull/11404
  - https://github.com/matrix-org/matrix-js-sdk/pull/3673
- https://github.com/matrix-org/matrix-react-sdk/pull/11353
  - https://github.com/matrix-org/matrix-js-sdk/pull/3647
- https://github.com/matrix-org/matrix-react-sdk/pull/11248
- https://github.com/matrix-org/matrix-react-sdk/pull/11182
- https://github.com/vector-im/element-web/pull/25923

### Screens

CreateRoomDialog

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * Allow creating public knock rooms ([\#11481](https://github.com/matrix-org/matrix-react-sdk/pull/11481)). Contributed by @charlynguyen.

### CODE REVIEW

**Comment by richvdh** [comment]
Sadly I think this is still blocked on an answer from the design team. @daniellekirkwood [wrote](https://github.com/matrix-org/matrix-react-sdk/pull/11464#issuecomment-1691902413):

> > You're happy for the setting to only appear for knockable rooms, and not (say) "public" rooms. (This seems particularly unintuitive to me)
> 
> Good point -- when someone selects Public it doesn't necessarily add it to the directory, right? Maybe this checkbox should appear there also... @nadonomy i would like support on that question...

@nadonomy / @daniellekirkwood can we get some clarity on why this checkbox only applies to knockable rooms?

File: src/components/views/elements/LabelledCheckbox.tsx [hunk 1]
```diff
@@ -29,11 +30,13 @@ interface IProps {
     disabled?: boolean;
     // The function to call when the value changes
     onChange(checked: boolean): void;
+    // Optional CSS class to apply to the label
```
Comment (richvdh) [inline src/components/views/elements/LabelledCheckbox.tsx:33]:
```suggestion
    // Optional CSS class to apply to the label
```

File: test/components/views/dialogs/CreateRoomDialog-test.tsx [hunk 2]
```diff
@@ -210,45 +210,72 @@ describe("<CreateRoomDialog />", () => {
     });
 
     describe("for a knock room", () => {
-        it("should not have the option to create a knock room", async () => {
-            jest.spyOn(SettingsStore, "getValue").mockReturnValue(false);
-            getComponent();
-            fireEvent.click(screen.getByLabelText("Room visibility"));
-
-            expect(screen.queryByRole("option", { name: "Ask to join" })).not.toBeInTheDocument();
+        describe("when disabling feature", () => {
```
Comment (richvdh) [inline test/components/views/dialogs/CreateRoomDialog-test.tsx:213]:
```suggestion
        describe("when feature is disabled", () => {
```

File: test/components/views/dialogs/CreateRoomDialog-test.tsx [hunk 3]
```diff
@@ -210,45 +210,72 @@ describe("<CreateRoomDialog />", () => {
...
+                jest.spyOn(SettingsStore, "getValue").mockReturnValue(false);
+                getComponent();
+                fireEvent.click(screen.getByLabelText("Room visibility"));
+                expect(screen.queryByRole("option", { name: "Ask to join" })).not.toBeInTheDocument();
+            });
         });
 
-        it("should create a knock room", async () => {
-            jest.spyOn(SettingsStore, "getValue").mockImplementation((setting) => setting === "feature_ask_to_join");
+        describe("when enabling feature", () => {
```
Comment (richvdh) [inline test/components/views/dialogs/CreateRoomDialog-test.tsx:222]:
```suggestion
        describe("when feature is enabled", () => {
```

File: test/components/views/dialogs/CreateRoomDialog-test.tsx [hunk 4]
```diff
@@ -210,45 +210,72 @@ describe("<CreateRoomDialog />", () => {
...
             const onFinished = jest.fn();
```
Comment (richvdh) [inline test/components/views/dialogs/CreateRoomDialog-test.tsx:223]: we either need to reset this mock after each test, or create a new one before each test.

**Review by richvdh — CHANGES_REQUESTED** [review]

**Comment by richvdh** [comment]
> Sadly I think this is still blocked on an answer from the design team. @daniellekirkwood [wrote](https://github.com/matrix-org/matrix-react-sdk/pull/11464#issuecomment-1691902413):
> 
> > > You're happy for the setting to only appear for knockable rooms, and not (say) "public" rooms. (This seems particularly unintuitive to me)
> > 
> > 
> > Good point -- when someone selects Public it doesn't necessarily add it to the directory, right? Maybe this checkbox should appear there also... @nadonomy i would like support on that question...
> 
> @nadonomy / @daniellekirkwood can we get some clarity on why this checkbox only applies to knockable rooms?

Design folks say: let's go with it as designed and implemented for now. We may return to this and clean it up in future.

**Comment by daniellekirkwood** [comment]
> We may return to this and clean it up in future

We will -- it's in the roadmap already :)

File: src/components/views/dialogs/CreateRoomDialog.tsx [hunk 5]
```diff
@@ -46,15 +46,25 @@ interface IProps {
...
+    // The room alias.
     alias: string;
+    // Indicates whether the details section is open.
     detailsOpen: boolean;
+    // Indicates whether federation is disabled for the room.
     noFederate: boolean;
+    // Indicates whether the room name is valid.
     nameIsValid: boolean;
+    // Indicates whether the user can change encryption settings for the room.
     canChangeEncryption: boolean;
```
Comment (richvdh) [inline src/components/views/dialogs/CreateRoomDialog.tsx:68]: thank you for adding all these! for future reference they should really be `/** ... */` -style comments for future interpretation by `tsdoc`.

File: src/components/views/dialogs/CreateRoomDialog.tsx [hunk 6]
```diff
@@ -45,15 +46,25 @@ interface IProps {
 }
 
 interface IState {
+    // The selected room join rule.
     joinRule: JoinRule;
-    isPublic: boolean;
+    // Indicates whether the knock room is public visible. Applies only for rooms with knock join rule.
```
Comment (richvdh) [inline src/components/views/dialogs/CreateRoomDialog.tsx:51]:
```suggestion
    /**
      * Indicates whether the created room should have public visibility (ie, it should be
      * shown in the public room list). Only applicable if `joinRule` == `JoinRule.Knock`.
      */
```

**Comment by richvdh** [comment]
@charlynguyen thanks for updating this! For future reference, please avoid force-pushing to git after a review has been done. Github makes it very hard to see what has changed since the previous review!

**Review by richvdh — APPROVED** [review]
thank you for your work on this!

**Pull request merged**